test/cors-gate.spec.js

```diff
@@ -0,0 +1,214 @@
+      .expect(403, done);
+  });
+
+  describe('options.allowSafe', function() {
+    it('should allow unspecified safe requests', function(done) {
+      this.app.use(corsGate({
+        strict: true,
+        allowSafe: true,
+        origin: 'http://localhost'
+      }));
+
+      this.app.get('/get', ok);
+
+      request(this.app)
+        .get('/get')
+        .expect(200, done);
+    });
+
+    it('should reject unspecified safe requests', function(done) {
+      this.app.use(corsGate({
+        strict: true,
+        allowSafe: false,
+        origin: 'http://localhost'
+      }));
+
+      this.app.get('/get', ok);
+
+      request(this.app)
+        .get('/get')
+        .expect(403, done);
+    });
+  });
+
+  describe('options.failure', function() {
+    it('should not be invoked for same-origin requests', function(done) {
+      this.app.use(corsGate({
+        allowSafe: false,
+        origin: 'http://localhost',
+        failure: function(req, res) {
+          res.status(403).end();
+        }
+      }));
+
+      this.app.post('/post', ok);
+
+      request(this.app)
+        .post('/post')
+        .set('origin', 'http://localhost')
+        .expect(200, done);
+    });
+
+
+    it('should be invoked for requests without origin', function(done) {
+      this.app.use(corsGate({
+        origin: 'http://localhost',
+        failure: function(req, res) {
+          res.status(204).end();
+        }
+      }));
+
+      this.app.post('/post', ok);
+
+      request(this.app)
+        .post('/post')
+        .expect(204, done);
+    });
+
+    it('should be invoked for requests from other origins', function(done) {
+      this.app.use(cors({
+        origin: ['http://localhost', 'http://localhost:8080']
+      }), corsGate({
+        origin: 'http://localhost',
+        failure: function(req, res) {
+          res.status(204).end();
+        }
+      }));
+
+      this.app.post('/post', ok);
+
+      request(this.app)
+        .post('/post')
+        .set('origin', 'https://mixmax.com')
+        .expect(204, done);
+    });
+  });
+
+  describe('originFallbackToReferrer', function() {
+    it('should patch origin', function(done) {
+      this.app.use(corsGate.originFallbackToReferrer());
+
+      this.app.get('/get', function(req, res) {
+        const correctOrigin = req.get('origin') === 'http://localhost';
+        res.status(correctOrigin ? 200 : 403).end();
+      });
+
+      request(this.app)
+        .get('/get')
+        .set('referer', 'http://localhost/home')
+        .expect(200, done);
+    });
+
+    it('should not overwrite origin', function(done) {
+      this.app.use(corsGate.originFallbackToReferrer());
+
+      this.app.get('/get', function(req, res) {
+        const correctOrigin = req.get('origin') === 'http://localhost';
+        res.status(correctOrigin ? 200 : 403).end();
+      });
+
+      request(this.app)
+        .get('/get')
+        .set('referer', 'http://localhost:8080/home')
```

Comment:
Good point. I just wanted to verify that the `Origin` header would be respected as canonical, rather than always pulling in data from the `Referer` when available.